rust/src/ldap/logger.rs

```diff
@@ -0,0 +1,351 @@
+            if !attr.attr_vals.is_empty() {
+                js.open_array("values")?;
+                for val in &attr.attr_vals {
+                    js.append_string_from_bytes(&val.0)?;
+                }
+                js.close()?;
+            }
+            js.close()?;
+        }
+        js.close()?;
+    }
+    js.close()?;
+    Ok(())
+}
+
+fn log_search_result_done(msg: &LdapResult, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("search_result_done")?;
+    log_ldap_result(msg, js)?;
+    js.close()?;
+    Ok(())
+}
+
+fn log_bind_response(msg: &BindResponse, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("bind_response")?;
+    log_ldap_result(&msg.result, js)?;
+    if let Some(creds) = &msg.server_sasl_creds {
+        js.set_hex("server_sasl_creds", creds)?;
+    };
+    js.close()?;
+    Ok(())
+}
+
+fn log_modify_response(msg: &ModifyResponse, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("modify_response")?;
+    log_ldap_result(&msg.result, js)?;
+    js.close()?;
+    Ok(())
+}
+
+fn log_add_response(msg: &LdapResult, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("add_response")?;
+    log_ldap_result(msg, js)?;
+    js.close()?;
+    Ok(())
+}
+
+fn log_del_response(msg: &LdapResult, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("del_response")?;
+    log_ldap_result(msg, js)?;
+    js.close()?;
+    Ok(())
+}
+
+fn log_mod_dn_response(msg: &LdapResult, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("mod_dn_response")?;
+    log_ldap_result(msg, js)?;
+    js.close()?;
+    Ok(())
+}
+
+fn log_compare_response(msg: &LdapResult, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("compare_response")?;
+    log_ldap_result(msg, js)?;
+    js.close()?;
+    Ok(())
+}
+
+fn log_extended_response(msg: &ExtendedResponse, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.open_object("extended_response")?;
+    log_ldap_result(&msg.result, js)?;
+    if let Some(name) = &msg.response_name {
+        js.set_string("name", &name.0)?;
+    }
+    if let Some(value) = &msg.response_value {
+        js.set_string_from_bytes("value", &value[..])?;
+    }
+    js.close()?;
+    Ok(())
+}
+
+fn log_intermediate_response(
+    msg: &IntermediateResponse, js: &mut JsonBuilder,
+) -> Result<(), JsonError> {
+    js.open_object("intermediate_response")?;
+    if let Some(name) = &msg.response_name {
+        js.set_string("name", &name.0)?;
+    }
+    if let Some(value) = &msg.response_value {
+        js.set_string_from_bytes("value", &value[..])?;
+    }
+    js.close()?;
+    Ok(())
+}
+
+fn log_ldap_result(msg: &LdapResult, js: &mut JsonBuilder) -> Result<(), JsonError> {
+    js.set_string("result_code", &msg.result_code.to_string())?;
+    js.set_string("matched_dn", &msg.matched_dn.0)?;
+    js.set_string("message", &msg.diagnostic_message.0)?;
```

Comment:
These 2 seem empty a lot. I wonder if we should only log them when there is data.